public/app/features/explore/state/query.test.ts

```diff
@@ -213,4 +226,95 @@ describe('reducer', () => {
         });
     });
   });
+
+  describe('caching', () => {
+    it('should add response to cache', async () => {
+      const store = configureStore({
+        ...(defaultInitialState as any),
+        explore: {
+          [ExploreId.left]: {
+            ...defaultInitialState.explore[ExploreId.left],
+            queryResponse: {
+              series: [{ name: 'test name' }] as DataFrame[],
+              state: LoadingState.Done,
+            } as PanelData,
+            absoluteRange: { from: 1621348027000, to: 1621348050000 },
+          },
+        },
+      });
+
+      await store.dispatch(addResultsToCache(ExploreId.left));
+
+      expect(store.getState().explore[ExploreId.left].cache).toEqual([
+        { key: 'from=1621348027000&to=1621348050000', value: { series: [{ name: 'test name' }], state: 'Done' } },
+      ]);
+    });
+
+    it('should not add response to cache if response is stil loading', async () => {
+      const store = configureStore({
+        ...(defaultInitialState as any),
+        explore: {
+          [ExploreId.left]: {
+            ...defaultInitialState.explore[ExploreId.left],
+            queryResponse: { series: [{ name: 'test name' }] as DataFrame[], state: LoadingState.Loading } as PanelData,
+            absoluteRange: { from: 1621348027000, to: 1621348050000 },
+          },
+        },
+      });
+
+      await store.dispatch(addResultsToCache(ExploreId.left));
+
+      expect(store.getState().explore[ExploreId.left].cache).toEqual([]);
+    });
+
+    it('should not add duplicate response to cache', async () => {
+      const store = configureStore({
+        ...(defaultInitialState as any),
+        explore: {
+          [ExploreId.left]: {
+            ...defaultInitialState.explore[ExploreId.left],
+            queryResponse: {
+              series: [{ name: 'test name' }] as DataFrame[],
+              state: LoadingState.Loading,
```

Comment:
Shouldn't this be `LoadingState.Done`? With `Loading` the cache is not updated because of the loading state (which is tested above).

Reply:
Checked the test is passing with `LoadingState.Done`. @davkal / @Elfo404 - does it make sense? I can change it.

Reply:
Yes, good catch! I've updated it.

Reply:
https://github.com/grafana/grafana/pull/34297/commits/a3af0a24d8c5159f2a25771cbcd9deecafa8dd5d